common/lending_iterator/lending_iterator.rs

```diff
@@ -0,0 +1,154 @@
+    where
+        Self: Sized,
+        B: 'static,
+        F: for<'a> FnMutHktHelper<Self::Item<'a>, B>,
+    {
+        Map::new(self, mapper)
+    }
+
+    fn filter_map<B, F>(self, mapper: F) -> FilterMap<Self, F, B>
+    where
+        Self: Sized,
+        B: Hkt + 'static,
+        F: for<'a> FnMutHktHelper<Self::Item<'a>, Option<B::HktSelf<'a>>>,
+    {
+        FilterMap::new(self, mapper)
+    }
+
+    fn into_iter(mut self) -> impl Iterator<Item = Self::Item<'static>>
+    where
+        Self: Sized,
+        for<'a> Self::Item<'a>: 'static,
+    {
+        iter::from_fn(move || unsafe {
+            // SAFETY: `Self::Item<'a>: 'static` implies that the item is independent from the iterator.
+            transmute::<Option<Self::Item<'_>>, Option<Self::Item<'static>>>(self.next())
+        })
+    }
+
+    fn collect<B>(self) -> B
+    where
+        Self: Sized,
+        for<'a> Self::Item<'a>: 'static,
+        B: FromIterator<Self::Item<'static>>,
+    {
+        self.into_iter().collect()
+    }
+
+    fn count(mut self) -> usize
+    where
+        Self: Sized,
+    {
+        let mut count = 0;
+        while self.next().is_some() {
+            count += 1;
+        }
+        count
+    }
+}
+
+pub trait Seekable<K: ?Sized>: LendingIterator {
+    fn seek(&mut self, key: &K);
+}
+
+pub struct Peekable<LI: LendingIterator> {
+    iter: LI,
+    item: Option<LI::Item<'static>>,
+}
+
+impl<LI: LendingIterator> Peekable<LI> {
+    pub fn new(iter: LI) -> Self {
+        Self { iter, item: None }
+    }
+
+    pub fn peek(&mut self) -> Option<&LI::Item<'_>> {
+        if self.item.is_none() {
+            self.item = unsafe {
+                // SAFETY: the stored item is only accessible while mutably borrowing this iterator.
+                // When the underlying iterator is advanced, the stored item is discarded.
+                transmute::<Option<LI::Item<'_>>, Option<LI::Item<'static>>>(self.iter.next())
+            };
+        }
+        unsafe {
+            // SAFETY: the item reference borrows this iterator mutably. This iterator cannot be advanced while it exists.
+            transmute::<Option<&LI::Item<'static>>, Option<&LI::Item<'_>>>(self.item.as_ref())
+        }
+    }
+}
+
+impl<LI: LendingIterator> LendingIterator for Peekable<LI> {
+    type Item<'a> = LI::Item<'a>;
+
+    fn next(&mut self) -> Option<Self::Item<'_>> {
+        match self.item.take() {
+            Some(item) => Some(unsafe {
+                // SAFETY: the item borrows this iterator mutably. This iterator cannot be advanced while it exists.
+                transmute(item)
+            }),
+            None => self.iter.next(),
+        }
+    }
+}
+
+impl<K: ?Sized, LI> Seekable<K> for Peekable<LI>
+where
+    LI: Seekable<K>,
+{
+    fn seek(&mut self, key: &K) {
+        if self.item.is_some() {
+            todo!()
+        }
+        self.item = None;
+        self.iter.seek(key)
```

Comment:
Let's establish the convention for all `seek` operations: Seeking into the previous iterator states is an illegal states. This is sort of a strict behaviour, but aligns with how we will be using seek() throughout TypeDB.